sway/config/output.c

```diff
@@ -821,19 +821,35 @@ static bool _spawn_swaybg(char **command) {
 			sway_log_errno(SWAY_ERROR, "execvp failed");
 			_exit(EXIT_FAILURE);
 		}
-		_exit(EXIT_SUCCESS);
+
+		// retry multiple times to test for failure;
+		// if process continues to run, it is considered to be successful
+		int swaybg_timeout = 10;
+		struct timespec sleep_time = { .tv_sec = 0, .tv_nsec = 1e6 };
+		int swaybg_status = 0;
+		while(swaybg_status == 0 && swaybg_timeout > 0) {
```

Comment:
We already setup a `wl_client` before-hand for swaybg (`config->swaybg_client`).

I'd rather not do any kind of PATH lookup, there is no function to do so in libc and each libc might have its own behavior.

What wlroots does is check for Xwayland at build-time, and then check that the Xwayland path configured at build-time exists at startup. It's not perfect (fails when the system-wide Xwayland is not available, but the user has configured one in PATH at a different location).

We could add startup notifications support to swaybg, and wait for that in Sway. Not sure it's entirely worth the trouble though.

Reply:
If we ignore the IPC case a bit, we could just improve the logging when execvp fails - right now it just prints "execvpe failed" with the errno, which is not particularly helpful. A message saying that background configuration was not applied as swaybg failed to execute would help give users a clue as to why their config is not taking effect.

Reply:
> If we must wait for something, it should preferably be a success indicator that triggers immediately in the good case, such as swaybg connected to the wayland display (would require sway to actively service the event loop while waiting) 

I don't really like my "solution" either, at the time I didn't see any other way to fix the described issue to get the error via `swaymsg`. I still think it is not really possible to avoid some waiting (without daemonizing or dry-running `swaybg` or similar), although I agree that it might be better to wait for success and not failure.
Serving the event loop in the mean time will probably increase code complexity by quite a bit since the IPC has to be kept alive until the success indicator is received (or not after some timeout).

> due to swaybg forking on success (which would be breaking change).

Why would be having some kind of `--daemonize` option and using it in `sway` for `swaybg` be breaking? Or did I misunderstand you? Wouldn't that kind of solve the whole issue?

> If we ignore the IPC case a bit, we could just improve the logging when execvp fails - right now it just prints "execvpe failed" with the errno, which is not particularly helpful. A message saying that background configuration was not applied as swaybg failed to execute would help give users a clue as to why their config is not taking effect.

I agree, I'll add a better log message which should definitely help when looking for the error.
From a user perspective, it might still be very confusing to have incorrect feedback from `swaymsg`.
But I'm also not sure if it is worth the trouble. :)

PS: I just noticed, that non-existing images also report `success` because `swaybg` will only log the error but not exit.

PPS: I just tried the current implementation on my desktop PC and it doesn't even work there anymore because the delay seems to be too short

Reply:
@kennylevinsen @emersion  I'd like to wrap up this PR, could you maybe give your opinion on introducing `--daemonize` or `--dry-run` (which still wouldn't detect any errors in the actual execution) for `swaybg`? With breaking change, do you mean that a specific minimum `swaybg` version would be required to contain the according flags?
Only other option I can think of would sending a "success" signal in `swaybg` and waiting for that in `sway` (with some kind of timeout) - but that would be "breaking" in the same way.

I don't really see another (pretty) solution to solve this - I can also keep it as-is (on the master branch) and only improve the logging if you don't think the approaches above could work and also have no other idea.

Reply:
As I said before, I wouldn't object to readiness notifications like the [`--ready-fd` option](https://github.com/swaywm/swaylock/pull/335) swaylock has. But not sure it's worth the trouble.

Reply:
Okay, thanks for the pointer. You're right - I also don't really think it is worth implementing since the flag could probably not be used for anything else and it would mainly catch if `swaybg` is installed (or can you think of something else causing that failure?) and the logging might be enough for that.

I pushed a new version which keeps the check in case forking fails and improves the logging.